base_layer/core/src/chain_storage/test/chain_backend.rs

```diff
@@ -0,0 +1,708 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+use crate::{
+    blocks::BlockHeader,
+    chain_storage::{
+        blockchain_database::BlockchainBackend,
+        db_transaction::{DbKey, DbKeyValuePair, DbValue, MetadataKey, MetadataValue},
+        lmdb_db::{
+            LMDB_DB_BLOCK_HASHES,
+            LMDB_DB_HEADERS,
+            LMDB_DB_HEADER_MMR_BASE_BACKEND,
+            LMDB_DB_HEADER_MMR_CP_BACKEND,
+            LMDB_DB_KERNELS,
+            LMDB_DB_KERNEL_MMR_BASE_BACKEND,
+            LMDB_DB_KERNEL_MMR_CP_BACKEND,
+            LMDB_DB_METADATA,
+            LMDB_DB_ORPHANS,
+            LMDB_DB_RANGE_PROOF_MMR_BASE_BACKEND,
+            LMDB_DB_RANGE_PROOF_MMR_CP_BACKEND,
+            LMDB_DB_STXOS,
+            LMDB_DB_UTXOS,
+            LMDB_DB_UTXO_MMR_BASE_BACKEND,
+            LMDB_DB_UTXO_MMR_CP_BACKEND,
+        },
+        DbTransaction,
+        LMDBDatabase,
+        MemoryDatabase,
+        MmrTree,
+    },
+    tari_amount::MicroTari,
+    test_utils::builders::{create_test_block, create_test_kernel, create_utxo},
+    tx,
+    types::HashDigest,
+};
+use tari_mmr::MutableMmr;
+use tari_storage::lmdb_store::{db, LMDBBuilder};
+use tari_test_utils::paths::create_random_database_path;
+use tari_utilities::{hex::Hex, Hashable};
+
+pub fn create_lmdb_database() -> LMDBDatabase<HashDigest> {
```

Comment:
I feel like this function should actually be a utility function in `lmdb_db.rs` with a signature

`pub fn create_lmdb_backend(path: &Path) -> LMDBDatabase<HashDigest> {...}`

which allows the node software to decide where the database files should reside.

Reply:
Change applied, but modified signature slightly so it also returns a result.